userspace/libpman/src/ringbuffer.c

```diff
@@ -201,13 +202,158 @@ int pman_finalize_ringbuf_array_after_loading()
 	return err;
 }
 
-static void *ringbuf__search_one_from_ring(struct ring *r, int ring_cnt)
+static inline void *ringbuf__get_first_ring_event(struct ring *r, int pos)
 {
-	int *len_ptr, len;
+	int *len_ptr = NULL;
+	int len = 0;
+	void *sample = NULL;
+
+	/* If the consumer reaches the producer update its position. */
+	if(g_state.cons_pos[pos] >= g_state.prod_pos[pos])
+	{
+		g_state.prod_pos[pos] = smp_load_acquire(r->producer_pos);
+		return NULL;
+	}
+
+	len_ptr = r->data + (g_state.cons_pos[pos] & r->mask);
+	len = smp_load_acquire(len_ptr);
+
+	/* The actual event is not yet committed */
+	if(len & BPF_RINGBUF_BUSY_BIT)
+	{
+		return NULL;
+	}
+
+	/* Save the size of the event if we need to increment the consumer */
+	g_state.last_event_size = roundup_len(len);
+
+	/* the sample is not discarded kernel side. */
+	if((len & BPF_RINGBUF_DISCARD_BIT) == 0)
+	{
+		sample = (void *)len_ptr + BPF_RINGBUF_HDR_SZ;
+	}
+
+	return sample;
+}
+
+static void ringbuf__consume_first_event(struct ring_buffer *rb, struct ppm_evt_hdr **event_ptr, int16_t *cpu_id)
+{
+	uint64_t min_ts = 0xffffffffffffffffLL;
+	struct ppm_evt_hdr *tmp_pointer = NULL;
+	int tmp_ring = -1;
+	unsigned long tmp_cons_increment = 0;
+
+	/* If the last consume operation was successful we can push the consumer position */
+	if(g_state.last_ring_read != -1)
+	{
+		struct ring *r = &(rb->rings[g_state.last_ring_read]);
+		g_state.cons_pos[g_state.last_ring_read] += g_state.last_event_size;
+		smp_store_release(r->consumer_pos, g_state.cons_pos[g_state.last_ring_read]);
+	}
+
+	for(uint16_t pos = 0; pos < rb->ring_cnt; pos++)
+	{
+		*event_ptr = ringbuf__get_first_ring_event(&rb->rings[pos], pos);
+
+		/* if NULL search for events in another buffer */
+		if(*event_ptr == NULL)
+		{
+			continue;
+		}
+
+		if((*event_ptr)->ts < min_ts)
+		{
+			min_ts = (*event_ptr)->ts;
+			tmp_pointer = *event_ptr;
+			tmp_ring = pos;
+			tmp_cons_increment = g_state.last_event_size;
+		}
+	}
+
+	*event_ptr = tmp_pointer;
+	*cpu_id = tmp_ring;
+	g_state.last_ring_read = tmp_ring;
+	g_state.last_event_size = tmp_cons_increment;
+}
+
+/* This API must be used if we want to get the first event according to its timestamp */
+void pman_consume_first_from_buffers(void **event_ptr, int16_t *cpu_id)
+{
+	ringbuf__consume_first_event(g_state.rb_manager, (struct ppm_evt_hdr **)event_ptr, cpu_id);
+}
```

Comment:
This is the most interesting part if there is something wrong here we are dead :skull: